### PR TITLE
#6077: Fix unet pcc issues

### PR DESCRIPTION
**Pull request by tapspatel**

### CODE REVIEW

**Comment by tapspatel** [comment]
Issue: https://github.com/tenstorrent-metal/tt-metal/issues/6077
Pipeline: https://github.com/tenstorrent-metal/tt-metal/actions/runs/8381339782

File: tests/ttnn/integration_tests/unet/test_ttnn_shallow_unet.py [hunk 1]
```diff
@@ -203,8 +203,8 @@ def custom_preprocessor(model, name, ttnn_module_args):
             ttnn_module_args.c8_3["conv_blocking_and_parallelization_config_override"] = {"act_block_h": 32}
 
             ttnn_module_args.output_layer["math_fidelity"] = ttnn.MathFidelity.LoFi
-            ttnn_module_args.output_layer["dtype"] = ttnn.bfloat8_b
-            ttnn_module_args.output_layer["weights_dtype"] = ttnn.bfloat8_b
+            ttnn_module_args.output_layer["dtype"] = ttnn.ttnn.bfloat16
```
Comment (dvartaniansTT) [inline tests/ttnn/integration_tests/unet/test_ttnn_shallow_unet.py:206]: how bad does PCC get with bfloat8_b?

File: tests/ttnn/integration_tests/unet/test_ttnn_shallow_unet.py [hunk 2]
```diff
@@ -321,87 +321,87 @@ def __init__(self):
         super(UNet, self).__init__()
         # Contracting Path
         self.c1 = nn.Conv2d(3, 16, kernel_size=3, padding=1)
-        self.b1 = nn.BatchNorm2d(16)
+        self.b1 = nn.BatchNorm2d(16, momentum=1)
```
Comment (dvartaniansTT) [inline tests/ttnn/integration_tests/unet/test_ttnn_shallow_unet.py:324]: why did we need to add momentum here? is it used at all in inference time?

File: tt_eager/tt_dnn/op_library/conv/kernels/bmm_tilize_untilize_all_weights_in_l1_single_output_block_width_dim.cpp [hunk 3]
```diff
@@ -155,10 +155,7 @@ void MAIN {
         uint32_t in1_index_inner_dim_h_offset = 0;
 
         #ifdef PACK_RELU
-        if constexpr ((tilize_in0 || spill) && !untilize_out) {
-            // for each output block we start we relu disabled so that intermediate results are not relu'd
-            PACK(( llk_pack_relu_config(ReluType::NO_RELU) ));
-        }
+        PACK(( llk_pack_relu_config(ReluType::NO_RELU) ));
```
Comment (dvartaniansTT) [inline tt_eager/tt_dnn/op_library/conv/kernels/bmm_tilize_untilize_all_weights_in_l1_single_output_block_width_dim.cpp:158]: where was relu being applied that it was not supposed to?

**Review by dvartaniansTT — COMMENTED** [review]
LGTM for the most part, just a few questions. Thanks for the hard work @tapspatel

**Comment by dvartaniansTT** [comment]
@tapspatel and @nsmithtt as far as I know the changes fix the PCC but mess up the perf. Do we want to merge it into main? Should we rename the files to have distinguish performant and accurate variants until the PCC fix is less hacky?

File: models/experimental/functional_unet/tt/ttnn_shallow_unet.py [hunk 4]
```diff
@@ -74,40 +74,86 @@ def __call__(self, device, input_tensor):
         profiler.tracy_message("c1")
         output_tensor = self.c1(input_tensor)
         output_tensor = self.c1_2(output_tensor)
-        save_c1_2_out = ttl.tensor.sharded_to_interleaved(output_tensor, ttnn.DRAM_MEMORY_CONFIG)
+        save_c1_2_out = ttl.tensor.sharded_to_interleaved(
```
[the remark this answers is not in the document]
Comment (tapspatel) [reply models/experimental/functional_unet/tt/ttnn_shallow_unet.py:77]: the downstream concat op requires both inputs to be the same type, otherwise it falls back to on host. This change allows it to execute on device.

File: models/experimental/functional_unet/tt/ttnn_shallow_unet.py [hunk 5]
```diff
@@ -74,40 +74,86 @@ def __call__(self, device, input_tensor):
...
+            output_tensor, ttnn.DRAM_MEMORY_CONFIG, output_dtype=ttl.tensor.DataType.BFLOAT16
+        )
         output_tensor = self.p4(output_tensor)
 
         profiler.tracy_message("bnc")
         output_tensor = unet_reshard(output_tensor, self.bnc.conv.input_sharded_memory_config, use_reshard=False)
         output_tensor = self.bnc(output_tensor)
         output_tensor = self.bnc_2(output_tensor)
 
+        # need to convert into interleaved, then back into sharded due to pcc issues
```
[the remark this answers is not in the document]
Comment (tapspatel) [reply models/experimental/functional_unet/tt/ttnn_shallow_unet.py:114]: yes its' expensive. not good for perf. there is a more fundamental bug which I reprod and filed here: https://github.com/tenstorrent-metal/tt-metal/issues/6661

File: tests/ttnn/integration_tests/unet/test_ttnn_shallow_unet.py (hunk 1 above)
Comment (tapspatel) [reply tests/ttnn/integration_tests/unet/test_ttnn_shallow_unet.py:206]: 0.95 from 0.97 on GS
0.97 from 0.99 on WH

File: tests/ttnn/integration_tests/unet/test_ttnn_shallow_unet.py (hunk 2 above)
Comment (tapspatel) [reply tests/ttnn/integration_tests/unet/test_ttnn_shallow_unet.py:324]: no, but the batch norm running var will change and exponentially get larger. This leads to the final output being inf values. To keep it steady, I added this var

File: tt_eager/tt_dnn/op_library/conv/kernels/bmm_tilize_untilize_all_weights_in_l1_single_output_block_width_dim.cpp (hunk 3 above)
Comment (tapspatel) [reply tt_eager/tt_dnn/op_library/conv/kernels/bmm_tilize_untilize_all_weights_in_l1_single_output_block_width_dim.cpp:158]: if we don’t have spill and reload and bias is enabled we don’t disable relu for the next block

**Comment by tapspatel** [comment]
I don't think we should merge yet, I will try to get all our other perf changes into tip and wait for https://github.com/tenstorrent-metal/tt-metal/issues/6661 to be fixed. I cleaned up the pcc issues for this PR for tracking.

File: tests/ttnn/integration_tests/unet/test_ttnn_shallow_unet.py [hunk 6]
```diff
@@ -503,88 +503,92 @@ def forward(self, x):
...
-    output_tensor = output_tensor.to(torch_input_tensor.dtype)
-
-    output_tensor = output_tensor[:, 0, :, :]
-    output_tensor = torch.reshape(
-        output_tensor, (output_tensor.shape[0], 1, output_tensor.shape[1], output_tensor.shape[2])
-    )
-    # todo: taps - Disable assert with pcc as pcc is really bad
-    # assert_with_pcc(torch_output_tensor, output_tensor, pcc=0.9999)
-    ttnn.close_device(device)
+    with torch.no_grad():
```
Comment (cfjchu) [inline tests/ttnn/integration_tests/unet/test_ttnn_shallow_unet.py:509]: I see that this test isn't under a `pytest`.. Does this test actually get run on CI? If not, we should add this so that this gets coverage.

**Review by cfjchu — APPROVED** [review]

File: tests/ttnn/integration_tests/unet/test_ttnn_shallow_unet.py (hunk 6 above)
Comment (tapspatel) [reply tests/ttnn/integration_tests/unet/test_ttnn_shallow_unet.py:510]: fixed
